[Flow] Add process_exporter component

spartan0x117 opened the pull request:
<!--

CONTRIBUTORS GUIDE: https://github.com/grafana/agent/blob/main/docs/developer/contributing.md#updating-the-changelog

If this is your first PR or you have not contributed in a while, we recommend
taking the time to review the guide. It gives helpful instructions for
contributors around things like how to update the changelog.

-->

#### PR Description
Adds a flow component for `process_exporter`.

#### Which issue(s) this PR fixes
Closes #2304 

#### Notes to the Reviewer
In the docs, the `process_matcher_config` was not embedding nicely in the above table, so I split it out, and one of the descriptions is quite lengthy so I converted it to html to get it to fit. Please let me know if there's a better way to do either/both of those things.

#### PR Checklist

- [X] CHANGELOG updated
- [X] Documentation added
- [X] Tests updated

erikbaranowski commented on docs/sources/flow/reference/components/prometheus.integration.process_exporter.md:
```diff
@@ -0,0 +1,92 @@
+### `process_matcher_config`
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`name`       | `string`                         | <pre><p>The name to use for identifying the process group name in the metric. By default, it uses the base path of the executable.</p><p>The following template variables are available:<ul><li> {{.Comm}}:      Basename of the original executable from /proc/\<pid\>/stat.<li> {{.ExeBase}}:   Basename of the executable from argv[0].<li> {{.ExeFull}}:  ...
```
maybe do a short description in the table for long ones and more details below in paragraphs. I think that's how others are done in this situation.

rfratto replied:
+1 try to keep the description in the table as short as possible so the table remains easy to view. 

Longer explanations can be done outside of the table.

erikbaranowski commented on docs/sources/flow/reference/components/prometheus.integration.process_exporter.md:
```diff
@@ -0,0 +1,92 @@
+`track_threads`     | `bool`                   | Report on per-threadname metrics as well. | `true` | no
+`gather_smaps`      | `bool`                   | Gather metrics from the smaps file, which contains proportional resident memory size. | `true` | no
+`recheck_on_scrape` | `bool`                   | Recheck process names on each scrape. | `true` | no
+`process_names`     | `process_matcher_config` | A collection of matching rules to use for deciding which processes to monitor. Each config can match multiple processes, which will be tracked as a single process "group."
```
I think this 1 goes in a blocks section and gets dropped off this table. Check out another doc with blocks at the top level for an example.

rfratto replied:
+1, we don't put blocks in the arguments section.

erikbaranowski commented on component/prometheus/integration/process_exporter/process_exporter.go:
```diff
@@ -0,0 +1,79 @@
+}
+
+// Config is the base config for this component.
+type Config struct {
```
I think we normally call this type Arguments by convention

erikbaranowski commented on component/prometheus/integration/process_exporter/process_exporter.go:
```diff
@@ -0,0 +1,79 @@
+
+// Config is the base config for this component.
+type Config struct {
+	ProcessExporter MatcherRules `river:"process_names,block,optional"`
```
Should this be `ProcessExporter []MatcherGroup`? This might change the river config a bit so I'll defer to what you think makes sense for this wrapper.

erikbaranowski reviewed (commented):
looking good, a few comments and considerations from me added

rfratto commented on component/prometheus/integration/process_exporter/process_exporter.go:
```diff
@@ -0,0 +1,79 @@
+	ExeRules     []string `river:"exe,attr,optional"`
+	CmdlineRules []string `river:"cmdline,attr,optional"`
+}
+type MatcherRules []MatcherGroup
```
It's probably not necessary to define this type here, and we can just use `[]MatcherGroup` directly.

rfratto commented on docs/sources/flow/reference/components/prometheus.integration.process_exporter.md:
````diff
@@ -0,0 +1,92 @@
+## Usage
+
+```river
+prometheus.integration.process_exporter "LABEL"{
````
```suggestion
prometheus.integration.process_exporter "LABEL" {
```

rfratto commented on docs/sources/flow/reference/components/prometheus.integration.process_exporter.md:
```diff
@@ -31,17 +31,33 @@ Name | Type | Description | Default | Required
 `exe`        | `list(string)`                   | A list of strings that match `argv[0]` for a process. If there are no slashes, only the basename of `argv[0]` needs to match. Otherwise, the name must be an exact match. For example, "postgres" may match any postgres binary, but `/usr/local/bin/postgres` will only match a postgres process with that exact path. If any of the strings match, the proc ...
 `cmdline`    | `list(string)`                   | A list of regular expressions applied to the `argv` of the process. Each regex here must match the corresponding argv for the process to be tracked. The first element that is matched is `argv[1]`. Regex captures are added to the .Matches map for use in the name. | | no
 
+The following template variables are available: 
```
```suggestion
The `name` argument can use the following template variables: 
```

spartan0x117 commented:
@rfratto @erikbaranowski Thanks for the feedback! I added some tests and updated the docs. Possibly pending https://github.com/grafana/agent/pull/3049 does anything else stick out?

rfratto commented on CHANGELOG.md:
```diff
@@ -49,6 +49,8 @@ Main (unreleased)
   - `prometheus.integration.consul` collects metrics from a consul installation
     (@captncraig)
   - `prometheus.integration.github` collects metrics from GitHub (@jcreixell)
+  - `prometheus.integration.process_exporter` aggregates and collects metrics by mining `/proc` (@spartan0x117)
```
"Scraping" is probably a more Prometheus-friendly term here :) 

```suggestion
  - `prometheus.integration.process_exporter` aggregates and collects metrics by scraping `/proc`. (@spartan0x117)
```

rfratto commented on docs/sources/flow/reference/components/prometheus.integration.process_exporter.md:
```diff
@@ -0,0 +1,108 @@
+
+# prometheus.integration.process_exporter
+The `prometheus.integration.process_exporter` component embeds
+[process_exporter](https://github.com/ncabatoff/process-exporter) for collecting process stats mined from `/proc`.
```
```suggestion
[process_exporter](https://github.com/ncabatoff/process-exporter) for collecting process stats from `/proc`.
```

rfratto replied:
Do we want to mention which operating systems this component works for?

rfratto commented on docs/sources/flow/reference/components/prometheus.integration.process_exporter.md:
```diff
@@ -0,0 +1,108 @@
+### process_names block
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`name`       | `string`                         | The name to use for identifying the process group name in the metric. By default, it uses the base path of the executable. (See below for template variable details). | `{{.ExeBase}}` | no
+`comm`       | `list(string)`                   | A list of strings that match the base executable name for a process, truncated to 15 characters. It is derived from reading the second field of `/proc/<pid>/stat`, stripped of parens. | | no
+`exe`        | `list(string)`                   | A list of strings that match `argv[0]` for a process. If there are no slashes, only the basename of `argv[0]` needs to match. Otherwise, the name must be an exact match. For example, "postgres" may match any postgres binary, but `/usr/local/bin/postgres` will only match a postgres process with that exact path. If any of the strings match, the proc ...
+`cmdline`    | `list(string)`                   | A list of regular expressions applied to the `argv` of the process. Each regex here must match the corresponding argv for the process to be tracked. The first element that is matched is `argv[1]`. Regex captures are added to the .Matches map for use in the name. | | no
```
The descriptions here are quite long. Can you use shorter sentences and move more detailed explanations to outside of the table?

spartan0x117 replied:
Will do!

rfratto commented on docs/sources/flow/reference/components/prometheus.integration.process_exporter.md:
```diff
@@ -0,0 +1,108 @@
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`procfs_path`       | `string`                 | procfs mountpoint. | `/proc` | no
+`track_children`    | `bool`                   | If a proc is tracked, also track any of it's children that are not part of their own group. | `true` | no
```
```suggestion
`track_children`    | `bool`                   | Whether to track a process' children. | `true` | no
```

rfratto commented on docs/sources/flow/reference/components/prometheus.integration.process_exporter.md:
```diff
@@ -0,0 +1,108 @@
+---- | ---- | ----------- | ------- | --------
+`procfs_path`       | `string`                 | procfs mountpoint. | `/proc` | no
+`track_children`    | `bool`                   | If a proc is tracked, also track any of it's children that are not part of their own group. | `true` | no
+`track_threads`     | `bool`                   | Report on per-threadname metrics as well. | `true` | no
```
```suggestion
`track_threads`     | `bool`                   | Report metrics for a process' individual threads.  | `true` | no
```

rfratto commented on docs/sources/flow/reference/components/prometheus.integration.process_exporter.md:
```diff
@@ -0,0 +1,108 @@
+`procfs_path`       | `string`                 | procfs mountpoint. | `/proc` | no
+`track_children`    | `bool`                   | If a proc is tracked, also track any of it's children that are not part of their own group. | `true` | no
+`track_threads`     | `bool`                   | Report on per-threadname metrics as well. | `true` | no
+`gather_smaps`      | `bool`                   | Gather metrics from the smaps file, which contains proportional resident memory size. | `true` | no
```
Can the info about proportional resident memory size be moved to a blurb outside of the table?

```suggestion
`gather_smaps`      | `bool`                   | Gather metrics from the smaps file for a process. | `true` | no
```

rfratto commented on docs/sources/flow/reference/components/prometheus.integration.process_exporter.md:
```diff
@@ -0,0 +1,108 @@
+
+Hierarchy        | Block      | Description | Required
+---------------- | ---------- | ----------- | --------
+process_names          | [process_names][]  | A collection of matching rules to use for deciding which processes to monitor. Each config can match multiple processes, which will be tracked as a single process "group." | no
```
This description is also pretty long, you can make this shorter like `Rules for deciding which processes to monitor.` and then move the longer description to the intro sentence of the `process_names block` section.

rfratto commented on docs/sources/flow/reference/components/prometheus.integration.process_exporter.md:
```diff
@@ -0,0 +1,108 @@
+
+Name      | Type                | Description
+--------- | ------------------- | -----------
+`targets` | `list(map(string))` | The targets that can be used to collect `process_exporter` metrics.
```
```suggestion
`targets` | `list(map(string))` | Targets that expose `process_exporter` metrics.
```

rfratto reviewed (commented):
Thanks @spartan0x117! The docs are coming along well, but I'd like to see the descriptions in tables cut down quite a bit. Big tables don't render well in the UI which is why I'd like to keep them short.

rfratto commented:
@spartan0x117 Can you bring in the latest changes to rename and rename the component to `prometheus.exporter.process` to align with #3049?

rfratto commented on component/prometheus/exporter/process/process.go:
```diff
@@ -0,0 +1,79 @@
+
+// Arguments configures the prometheus.exporter.process component
+type Arguments struct {
+	ProcessExporter []MatcherGroup `river:"process_names,block,optional"`
```
Typically blocks should have singular names, since it's weird to say 

```river 
rules {
  ... 
}

rules { 
  ...
}
```

vs 

```
rule { 

}

rule {

}
```

Is `process_names` right here? Or should it be given a singular name?

spartan0x117 replied:
This is kept over from the old config. Since `process_names` ultimately can actually match based on a combination of (1) the name in `/proc/<pid>/stat`, (2) the base name or path to the file, or (3) a regex of cmdline arguments, maybe something more generic and singular like `process_matcher` or `matcher` would be more appropriate?

So an example config would look like:
```
prometheus.exporter.process "example_process" {
    process_matcher {
        name = "non_flow_agents"
        comm = ["grafana-agent"]
        cmdline = ["*config.file*"]
    }

    process_matcher {
        name = "firefox"
        comm = ["firefox-bin"]
    }
}
```

rfratto replied:
`matcher` seems best, I think process is probably implied given the component it's in.

rfratto commented on docs/sources/flow/reference/components/prometheus.exporter.process.md:
```diff
@@ -0,0 +1,116 @@
+# Ideally, in the future, we can fix the overflow issue with css rather than
+# injecting special characters.
+
+title: prometheus.​integration.​process_exporter
```
```suggestion
title: prometheus.​exporter.process
```

rfratto commented on docs/sources/flow/reference/components/prometheus.exporter.process.md:
```diff
@@ -0,0 +1,116 @@
+title: prometheus.​integration.​process_exporter
+---
+
+# prometheus.integration.process_exporter
```
```suggestion
# prometheus.exporter.process
```

rfratto commented on docs/sources/flow/reference/components/prometheus.exporter.process.md:
```diff
@@ -0,0 +1,116 @@
+---
+
+# prometheus.integration.process_exporter
+The `prometheus.integration.process_exporter` component embeds
```
```suggestion
The `prometheus.exporter.process` component embeds
```

rfratto commented on docs/sources/flow/reference/components/prometheus.exporter.process.md:
````diff
@@ -0,0 +1,116 @@
+## Usage
+
+```river
+prometheus.integration.process_exporter "LABEL" {
````
```suggestion
prometheus.exporter.process "LABEL" {
```

rfratto commented on docs/sources/flow/reference/components/prometheus.exporter.process.md:
```diff
@@ -0,0 +1,116 @@
+`recheck_on_scrape` | `bool`                   | Recheck process names on each scrape. | `true` | no
+
+## Blocks
+The following blocks are supported inside the definition of `prometheus.integration.process_exporter`:
```
```suggestion
The following blocks are supported inside the definition of `prometheus.exporter.process`:
```

rfratto commented on docs/sources/flow/reference/components/prometheus.exporter.process.md:
```diff
@@ -0,0 +1,116 @@
+
+## Component health
+
+`prometheus.integration.process_exporter` is only reported as unhealthy if given
```
```suggestion
`prometheus.exporter.process` is only reported as unhealthy if given
```

rfratto commented on docs/sources/flow/reference/components/prometheus.exporter.process.md:
```diff
@@ -0,0 +1,116 @@
+
+## Debug information
+
+`prometheus.integration.process_exporter` does not expose any component-specific
```
```suggestion
`prometheus.exporter.process` does not expose any component-specific
```

rfratto commented on docs/sources/flow/reference/components/prometheus.exporter.process.md:
```diff
@@ -0,0 +1,116 @@
+
+## Debug metrics
+
+`prometheus.integration.process_exporter` does not expose any component-specific
```
```suggestion
`prometheus.exporter.process` does not expose any component-specific
```

rfratto commented on docs/sources/flow/reference/components/prometheus.exporter.process.md:
```diff
@@ -0,0 +1,116 @@
+## Example
+
+This example uses a [`prometheus.scrape` component][scrape] to collect metrics
+from `prometheus.integration.process_exporter`:
```
```suggestion
from `prometheus.exporter.process`:
```

rfratto commented on docs/sources/flow/reference/components/prometheus.exporter.process.md:
````diff
@@ -0,0 +1,116 @@
+from `prometheus.integration.process_exporter`:
+
+```river
+prometheus.integration.process_exporter "example" {
````
```suggestion
prometheus.exporter.process "example" {
```

rfratto commented on docs/sources/flow/reference/components/prometheus.exporter.process.md:
```diff
@@ -0,0 +1,116 @@
+
+// Configure a prometheus.scrape component to collect process_exporter metrics.
+prometheus.scrape "demo" {
+  targets    = prometheus.integration.process_exporter.example.targets
```
```suggestion
  targets    = prometheus.exporter.process.example.targets
```

rfratto reviewed (approved):
LGTM, though I'm not sure about the block name. I'll leave it up to you what to call it though :)

rfratto reviewed (approved):
🎉

spartan0x117 commented:
It took a little while, but we got there :smile: Thanks for your help!
